### ISSUE TITLE
`Found unparsable section` when using Snowflake dialect

### ISSUE THREAD

**Issue by Xumeiquer** [sqlfluff/sqlfluff#5366]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I am using `SYSTEM$CURRENT_USER_TASK_NAME()` system function documented in [here](https://docs.snowflake.com/en/sql-reference/functions/system_current_user_task_name).

When linting my SQL queries `sqlfluff` raises an error `Found unparsable section: "()`.

The command used was: `sqlfluff lint --dialect snowflake sql.sql`

### Expected Behaviour

SQLFluff parse/lint the SQL expression without errors.

### Observed Behaviour

```
sqlfluff lint --dialect snowflake test.sql
== [test.sql] FAIL
L:   1 | P:   1 | AM04 | Query produces an unknown number of result columns.
                       | [ambiguous.column_count]
L:   2 | P:   6 | RF05 | Do not use special characters in identifiers.
                       | [references.special_chars]
L:   2 | P:  35 |  PRS | Line 2, Position 35: Found unparsable section: '()'
L:   2 | P:  37 | CV06 | Statements must end with a semi-colon.
                       | [convention.terminator]
WARNING: Parsing errors found and dialect is set to 'snowflake'. Have you configured your dialect correctly?
```

### How to reproduce

```sql
$ cat test.sql
select *
from SYSTEM$CURRENT_USER_TASK_NAME();
```

```bash
$ sqlfluff lint --dialect snowflake test.sql
```

### Dialect

snowflake

### Version

2.3.5

### Configuration

```ini
[sqlfluff]
dialect = snowflake
max_line_length = 80
# exclude some whitespace rules
# see https://docs.sqlfluff.com/en/stable/rules.html
exclude_rules = LT01, LT12, LT13

[sqlfluff:rules:convention.terminator]
multiline_newline = True
require_final_semicolon = True

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = lower
[sqlfluff:rules:capitalisation.identifiers]
capitalisation_policy = lower
[sqlfluff:rules:capitalisation.functions]
extended_capitalisation_policy = lower
[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = lower
[sqlfluff:rules:capitalisation.types]
extended_capitalisation_policy = lower
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by greg-finley** [comment]
Your query seems invalid in Snowflake per my own testing (it's just `select SYSTEM$CURRENT_USER_TASK_NAME()` not `select * from SYSTEM$CURRENT_USER_TASK_NAME()`):

<img width="1378" alt="Screenshot 2023-10-31 at 1 22 35 PM" src="https://github.com/sqlfluff/sqlfluff/assets/28961086/4b622366-1600-4e61-ba2c-013e081316d2">
<img width="1364" alt="Screenshot 2023-10-31 at 1 22 24 PM" src="https://github.com/sqlfluff/sqlfluff/assets/28961086/450829ab-618e-4b68-b4f7-7401c2ef34a2">

The valid one `select SYSTEM$CURRENT_USER_TASK_NAME ();` already parses: https://online.sqlfluff.com/fluffed?sql=H4sIAC9iQWUC_ytOzUlNLlEIjgwOcfVVcQ4NCnL1C4kPDXYNig9xDPaO93P0dVXQ0LTm5QIA-C5IsyoAAAA%3D&dialect=snowflake
